Microsoft.Azure.Cosmos/src/Resource/Container/Container.cs

```
@@ -1325,6 +1324,25 @@ public abstract class Container
             string processorName,
             ChangesHandler<T> onChangesDelegate);
 
+        /// <summary>
+        /// Initializes a <see cref="GetChangeFeedProcessorBuilderWithAllVersionsAndDeletes"/> for change feed processing with all versions and deletes.
+        /// </summary>
+        /// <typeparam name="T">Document type</typeparam>
+        /// <param name="processorName">A name that identifies the Processor and the particular work it will do.</param>
+        /// <param name="onChangesHandler">Delegate to receive all changes and deletes</param>
+        /// <returns>An instance of <see cref="ChangeFeedProcessorBuilder"/></returns>
+#if PREVIEW
+        public
+#else
+        internal
+#endif
+        virtual ChangeFeedProcessorBuilder GetChangeFeedProcessorBuilderWithAllVersionsAndDeletes<T>(
```

The only design-level question I have is: If a customer starts a CFP with AllVersionsAndDeletes and afterwards changes it to LatestVersion (or viceversa) should it pick up the lease containuation? is that something we want to support? What was done in Java? cc @jcocchi

---

Could it be that they are using the mode as part of the lease `id`? Can Abhijeet confirm this?

---

> Could it be that they are using the mode as part of the lease `id`? Can Abhijeet confirm this?

@jeet1995 can you weigh in. Thanks.

---

Otherwise how can that combination/condition will be detected? 

> CFP will not return documents if the mode is switched and also no exceptions 
Combination is very interesting. Won't CX interpret it as no-changes and CRI to us?

---

If the mode is added to the `id` of the leases, then each mode will have different set of leases (similar to processorName/leasePrefix) and they won't mix. But we should be consistent with what Java did

---

Thanks @jeet1995 for the details. I think the current code in this PR aligns with this behavior.

The point, which now might be a question for @jcocchi, is that the Continuation from one mode to the other is in theory valid, because it's just the LSN. The service won't fail or reject the request, but the behavior as described above, might be weird. 

Mainly because when switching inbetween modes, there might be scenarios that don't seem compatible. For example, if running an Incremental CFP, the Continuation might be an LSN older than the Full Fidelity Retention. If that CFP switches to AllVersionsAndDeletes, then the Continuation passed will be outside of retention.

In a nutshell the question is, do we want to support customers changing the mode? Do we want each mode to track different progress (for example, if we add the mode to the `id`, then each mode would create different set of leases and they would not mix) or do we want to throw an exception (the `mode` should be part of the lease and we should verify it matches with the CFP Builder mode).

---

Few comments:
- `AllVersionsAndDeletes` shouldn't support `startFromBeginning`, if this doesn't error currently it should definitely error. Let's track this as a future update for Java and verify the behavior in .NET.
- In the future, I want to leave the option open to support switching modes with the same leases, so I don't want to add `mode` to the `id`. Let's error on this scenario for now until we can test all edge cases. If the `mode` doesn't match what's stored in the lease then throw an exception.
    - @ealsur Reading an LSN outside of the retention could happen when switching modes as you mentioned, but it can also happen when reading old leases from `AllVersionsAndDeletes` mode itself. Switching modes may make it more likely for changes to be outside of retention, but deciding if a LSN is outside of the retention should happen in the backend and doesn't need to be a blocker for SDK design IMO. 
- @jeet1995 I'm confused about scenario 1. Sounds like you ingested 10 documents, read the 10 documents in `AllVersionsAndDeletes` mode then switched to `LatestVersion` mode. There were no new documents ingested after that so why would `LatestVersion` pick up anything? The `startFromBeginning` key only applies for the first run, if there is existing state in the leases then that takes precedence ([reference doc](https://azure.github.io/azure-cosmosdb-java/3.0.0/com/azure/data/cosmos/ChangeFeedProcessorOptions.html#startFromBeginning--)). Sounds like this scenario is functionally the same as scenario 2 for Java. @philipthomas-MSFT let's verify what happens in .NET.

---

@jcocchi - my bad - there is no prior test coverage for the lease sharing scenario therefore my assumption was a mode change shouldn't support a lease sharing scenario at least out of the box / the API ref doesn't call it out / `AllVersionAndDeletes` doesn't error out either when `StartFrom` is `BEGINNING` in the lease - so different modes don't have contextual knowledge of the same lease as of yet. Once I ran through the scenario, the lease is indeed shared with continuation being picked up in-line with the doc. I can validate the other scenario as well.

---

> `startFromBeginning`: Will it never be supported ever? 

SDK once shipped is immutable. So one option is to error out from service => can evolve as service gets enriched more. True experience wise it needs one extra NW call for it to fail, seems a reasonable trade-off. 

> In the future, I want to leave the option open to support switching modes with the same leases, so I don't want to add mode to the id

Switching modes practically won't be atomic right? ex: multi-instance processor and the change needs to roll-through all nodes, otherwise the behavior will be non-deterministic. 

One option/alternative is to treat them as migrations. I.e. create a new lease container with the new mode with current LSN state. 

@jcocchi thoughts?

---

`startFromBeginning` should error from the service, and I tested that it does on the .NET Pull model. Not sure why CFP doesn't see it as invalid. Maybe CFP is only looking at LSNs, and the LSN was still in the retention window for this case? By design `startFromBeginning` means "beginning of container" `AllVersionsAndDeletes` max retention is 7 days or 30 days, so `startFromBeginning` could only work for new containers in that mode.

@kirankumarkolli agree switching modes needs design work and testing. Let's error for now without closing the opportunity to add support for this in the future. My only comment was adding `mode` to the `id` might make it difficult to support the scenario later.

---

With CFP, the processor might attempt NW interaction only after its started, i.e. in background. So error will be observed by the application code (not-straight forward for troubleshooting). Can it be treated as fatal failure and stop the processor? @ealsur thoughts? 

On `switching modes`: Even to detect it the lease needs to have the previous mode context right? 
We can include the mode part of the lease and later in logic we can ignore if needed (logic change new CFP/SDK version refresh is needed by application)

---

The only way is if we wired the detection of the mode change through the StartAsync operation. That is the only way you can stop the CFP from starting. Once started, everything is background, the processing cannot be stopped.

If we throw exceptions, if the customer has Notifications API wired, they would see them, but that's all

---

That's right, even mode validation needs interaction with lead-container => NW interaction. May be StartAsync can be delayed till mode validation. 

`startfrombegining` in non-recoverable/fatal in-theory. May be we can get started with NotificationsAPI for now and later revisit based on CRI support-load?

---

@jcocchi / @philipthomas-MSFT - stopping `LatestVersion` CFP after an arbitrary no. of documents have been processed and continuing with `AllVersionsAndDeletes` CFP picks up from the continuation of the 'shared' lease in the Java SDK. 

For ref - https://github.com/jeet1995/azure-sdk-for-java/pull/8

---

Hi. I am going to resolve this, but move this conversation since we are not blocking this PR. We will continue this conversation on a new issue once I can gather more information and grow my understanding. Thanks.

---

Thanks for verifying in Java @jeet1995! 

I think that is the desired long-term behavior, but based on this conversation both .NET and Java should throw when the change feed mode is switched and the leases are the same for the time being. We can cover it offline, and it's non-blocking as Philip mentioned.

---

Compiling all the feedback in this discussion and transferring it [here](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/4308) along with some clarifying questions. Once I have a solid understanding and alignment on the direction, I will finalize, implement and deliver. Thanks. @kirankumarkolli @ealsur @jeet1995 @jcocchi